File: src/main/java/akostenko/aicars/math/Polar.java
package akostenko.aicars.math;

import static java.lang.StrictMath.PI;
import static java.lang.StrictMath.abs;
import static java.lang.StrictMath.asin;
import static java.lang.StrictMath.cos;
import static java.lang.StrictMath.sin;
import static java.lang.StrictMath.sqrt;

public class Polar implements Vector {
    public static final Polar ZERO = new Polar(0,0);

    /** length of vector, always >= 0 */
    public final double r;
    /** direction, <i>radians</i>, measured from X axis towards Y axis, always within [0, 2*PI) */
    public final double d;

    public Polar(double r, double d) {
        if (r < 0) {
            d += PI;
        }
        this.r = r;
        while (d < 0) d += 2*PI;
        while (d >= 2*PI) d -= 2*PI;
        this.d = d;
    }

    @Override
    public Polar toPolar() {
        return this;
    }

    @Override
    public Decart toDecart() {
        return new Decart(r * cos(d), r * sin(d));
    }

    @Override
    public Polar plus(Vector v) {
        return toDecart().plus(v.toDecart()).toPolar();
    }

    @Override
    public Polar minus(Vector v) {
        return plus(v.negative());
    }

    @Override
    public Polar negative() {
        return multi(-1);
    }

    @Override
    public Polar multi(double k) {
        if (k == 0) {
            return ZERO;
        } else {
            return new Polar(r * abs(k), d + (k > 0 ? 0 : PI));
        }
    }

    @Override
    public Polar div(double k) {
        return multi(1./k);
    }

    @Override
    public Polar rotate(double radians) {
        return new Polar(r, d+radians);
    }

    @Override
    public double module() {
        return r;
    }

    @Override
    public double moduleSqr() {
        return r*r;
    }

    @Override
    public double dot(Vector v) {
        Polar b = v.toPolar();
        return r*b.r*cos(d - b.d);
    }

    @Override
    public double cross(Vector v) {
        Polar b = v.toPolar();
        return r * b.r * sin(b.d - d);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;

        Polar polar;
        if (o instanceof Vector) {
            polar = ((Vector) o).toPolar();
        } else return false;

        return (polar.r - r < PRECISION) && (polar.d - d < PRECISION);
    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        temp = Double.doubleToLongBits(r);
        result = (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(d);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        return result;
    }

    @Override
    public String toString() {
        return "Polar{" +
                "r=" + r +
                ", d=" + d +
                '}';
    }
}

File: src/main/kotlin/akostenko/aicars/math/Polar.kt
package akostenko.aicars.math

import java.lang.StrictMath.PI
import java.lang.StrictMath.abs
import java.lang.StrictMath.cos
import java.lang.StrictMath.sin

class Polar(
        /** length of vector, always >= 0  */
        r: Double, d: Double) : Vector {
    /** direction, *radians*, measured from X axis towards Y axis, always within [-PI, +PI]  */
    val r: Double
    val d: Double
    // TODO reuse this into Decart.polar and vise versa
    private val decart : Decart by lazy { Decart(r * cos(d), r * sin(d)) }

    init {
        var d = d
        if (r < 0) {
            d += PI
        }
        while (d < -PI) d += 2 * PI
        while (d > PI) d -= 2 * PI
        this.d = d
        this.r = abs(r)
    }

    override fun toPolar() = this

    override fun toDecart() = decart

    override fun plus(v: Vector) = toDecart().plus(v.toDecart()).toPolar()

    override fun minus(v: Vector) = plus(-v)

    override fun unaryMinus() = times(-1.0)

    override fun times(k: Double): Polar {
        if (k == 0.0) {
            return ZERO
        } else {
            return Polar(r * abs(k), d + if (k > 0) 0.0 else StrictMath.PI)
        }
    }

    override fun div(k: Double) = times(1.0 / k)

    override fun rotate(radians: Double) = Polar(r, d + radians)

    override fun module() = r

    override fun moduleSqr() = r * r

    override fun dot(v: Vector): Double {
        val other = v.toPolar()
        return r * other.r * cos(d - other.d)
    }

    override fun cross(v: Vector): Double {
        val other = v.toPolar()
        return r * other.r * sin(other.d - d)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true

        val polar: Polar
        if (other is Vector) {
            polar = other.toPolar()
            return abs(polar.r - r) < Vector.PRECISION && abs(polar.d - d) < Vector.PRECISION
        } else return false
    }

    override fun hashCode(): Int {
        var result = r.hashCode()
        result = 31 * result + d.hashCode()
        return result
    }

    override fun toString(): String = "Polar(r=$r, d=$d)"

    companion object {
        val ZERO = Polar(0.0, 0.0)
    }
}
